nibabel/brainvoyager/bv.py

```diff
@@ -0,0 +1,952 @@
+       outside the current scope (see below).
+
+    Returns
+    -------
+    hdr_dict_new : OrderedDict
+        An updated version hdr_dict correcting effects of changed nested and
+        conditional fields.
+    """
+    for name, format, def_or_name in hdr_dict_proto:
+        # handle nested loop fields
+        if isinstance(format, tuple):
+            # calculate the change of array length and the new array length
+            if def_or_name in hdr_dict_old:
+                delta_values = hdr_dict_new[def_or_name] - \
+                    hdr_dict_old[def_or_name]
+                n_values = hdr_dict_new[def_or_name]
+            else:
+                delta_values = parent_new[def_or_name] - \
+                    parent_old[def_or_name]
+                n_values = parent_new[def_or_name]
+            if delta_values > 0:  # add nested loops
+                for i in range(delta_values):
+                    hdr_dict_new[name].append(_proto2default(format, hdr_dict_new))
+            elif delta_values < 0:  # remove nested loops
+                for i in range(abs(delta_values)):
+                    hdr_dict_new[name].pop()
+            # loop over nested fields
+            for i in range(n_values):
+                update_BV_header(format, hdr_dict_old[name][i],
+                                 hdr_dict_new[name][i], hdr_dict_old,
+                                 hdr_dict_new)
+    return hdr_dict_new
+
+
+def _proto2default(proto, parent_default_hdr=None):
+    """Helper for creating a BV header OrderedDict with default parameters.
+
+    Create an OrderedDict that contains keys with the header fields, and
+    default values.
+
+    See :func:`parse_BV_header` for description of `proto` format.
+    """
+    default_hdr = OrderedDict()
+    for name, format, def_or_name in proto:
+        if isinstance(format, tuple):
+            value = []
+            # check the length of the array to expect
+            if def_or_name in default_hdr:
+                n_values = default_hdr[def_or_name]
+            else:
+                n_values = parent_default_hdr[def_or_name]
+            for i in range(n_values):
+                value.append(_proto2default(format, default_hdr))
+            default_hdr[name] = value
+        # handle conditional fields
+        elif isinstance(def_or_name, tuple):
+            default_hdr[name] = def_or_name[0]
+        else:
+            default_hdr[name] = def_or_name
+    return default_hdr
+
+
+def combine_st(st_array, inv=False):
+    """Combine spatial transformation matrices.
+
+    This recursive function returns the dot product of all spatial
+    transformation matrices given in st_array for applying them in one go.
+    The order of multiplication follow the order in the given array.
+
+    Parameters
+    ----------
+    st_array: array
+        array filled with transformation matrices of shape (4, 4)
+
+    inv: boolean
+        Set to true to invert the transformation matrices before
```

Comment:
OK, I am waiting for a connection flight right now that was delayed 5 hours and I am already jetlagged, so I should definitely not make up my mind about spatial transformation matrices, BUT:
What I want here is to compute a transformation matrix from the orientation of the data in the BV file to their native orientation by reversing all the transformations that were performed on the data. These transformations are stored (e.g.) in the VMR file header in an ordered fashion. So I guess if taking the dot product from left to right gets me there, this is what I want.
But again, I will better try this out with a clear mind :)